File: electron/lib/paths.js
'use strict';

/**
 * Filesystem locations for the desktop app.
 *
 * Bundled resources (Modelfiles, asset manifest) live under
 * process.resourcesPath when packaged, or the repo root during `npm start`.
 * Per-install state (generated env file, cached downloads, first-run marker)
 * lives under Electron's userData dir so it survives app updates.
 */

const path = require('path');
const { app } = require('electron');

function resourcesRoot() {
  // Packaged: extraResources are copied straight under resourcesPath.
  // Dev (`electron .`): fall back to the repo root two levels up from lib/.
  return app.isPackaged
    ? process.resourcesPath
    : path.join(__dirname, '..', '..');
}

function modelfilePath(name) {
  // name is "Modelfile" or "Modelfile.embeddings".
  return path.join(resourcesRoot(), name);
}

function assetsManifestPath() {
  return path.join(resourcesRoot(), 'assets.manifest.json');
}

function userDataDir() {
  return app.getPath('userData');
}

/** Cache dir for large downloaded assets (image tar, dump, project_data). */
function assetsDir() {
  return path.join(userDataDir(), 'assets');
}

/** Extracted project_data corpus, bind-mounted read-only into the API. */
function projectDataDir() {
  return path.join(userDataDir(), 'project_data');
}

/** Dir holding the Neo4j dump for the offline load (mounted as /snapshot). */
function snapshotDir() {
  return path.join(userDataDir(), 'snapshot');
}

/** Per-install compose env file (random Neo4j password, host paths). */
function envFilePath() {
  return path.join(userDataDir(), 'desktop.env');
}

/** Marker written once the guided first-run provisioning has fully succeeded. */
function firstRunMarkerPath() {
  return path.join(userDataDir(), 'first-run-complete.json');
}

/**
 * Rotating log files: electron.log (shell + child output) and api.log (handed to
 * the API as LOG_DIR). The folder a user sends when reporting a problem.
 */
function logsDir() {
  return path.join(userDataDir(), 'logs');
}

// --- Native runtime locations (decontainerized) ---------------------------
// The bundled Neo4j server, its JRE, and the frozen API bundle are unpacked
// under userData by the first-run provisioning (Workstream C/D). The build
// pipeline (Workstream E) must produce these layouts:
//   <userData>/neo4j/bin/neo4j(.bat) + neo4j-admin(.bat), data/, conf/
//   <userData>/jre/            (bundled JRE; used as JAVA_HOME)
//   <userData>/api/emb-api(.exe) + _internal/   (PyInstaller one-dir bundle)

const _WIN = process.platform === 'win32';

/** Root of the unpacked native Neo4j server. */
function neo4jHomeDir() {
  return path.join(userDataDir(), 'neo4j');
}

function neo4jBinDir() {
  return path.join(neo4jHomeDir(), 'bin');
}

/** `neo4j` launcher — run with `console` to start the server in-process. */
function neo4jConsolePath() {
  return path.join(neo4jBinDir(), _WIN ? 'neo4j.bat' : 'neo4j');
}

/** `neo4j-admin` launcher — used for the offline snapshot load. */
function neo4jAdminPath() {
  return path.join(neo4jBinDir(), _WIN ? 'neo4j-admin.bat' : 'neo4j-admin');
}

/** Bundled JRE home; exported to child processes as JAVA_HOME for Neo4j. */
function jreHomeDir() {
  return path.join(userDataDir(), 'jre');
}

/** Unpacked PyInstaller one-dir bundle of the API. */
function apiDir() {
  return path.join(userDataDir(), 'api');
}

/** The frozen API executable inside apiDir(). */
function apiExePath() {
  return path.join(apiDir(), _WIN ? 'emb-api.exe' : 'emb-api');
}

module.exports = {
  resourcesRoot,
  modelfilePath,
  assetsManifestPath,
  userDataDir,
  assetsDir,
  projectDataDir,
  snapshotDir,
  envFilePath,
  firstRunMarkerPath,
  logsDir,
  neo4jHomeDir,
  neo4jBinDir,
  neo4jConsolePath,
  neo4jAdminPath,
  jreHomeDir,
  apiDir,
  apiExePath,
};

File: src/core/paths.py
"""Resource paths that must resolve both from source and inside a frozen bundle.

The decontainerized app runs two ways: from source under ``uvicorn`` (a venv),
and — once packaged — as a PyInstaller executable. Files the app reads at runtime
(the Ollama ``Modelfile`` variants, the static web UI) live at repo-relative
paths that only work from source. These helpers resolve them correctly in both
cases so nothing depends on the current working directory.

Packaging note: a frozen build must bundle these resources so they land under
:func:`resource_root` — e.g. PyInstaller ``--add-data`` for ``Modelfile``,
``Modelfile.embeddings`` (at the root) and ``src/static`` (preserving that
subpath). See ``docs/DECONTAINERIZE_PLAN.md`` (Workstream E).
"""

import sys
from pathlib import Path


def resource_root() -> Path:
    """Directory under which bundled runtime resources live.

    - From source: the repository root (two levels up from ``src/core``).
    - Frozen (PyInstaller): ``sys._MEIPASS`` for a one-file build, else the
      directory holding the executable for a one-dir build.
    """
    if getattr(sys, "frozen", False):
        meipass = getattr(sys, "_MEIPASS", None)
        return Path(meipass) if meipass else Path(sys.executable).resolve().parent
    return Path(__file__).resolve().parents[2]


def modelfile_path(name: str) -> Path:
    """Absolute path to a Modelfile (``Modelfile`` / ``Modelfile.embeddings``)."""
    return resource_root() / name


def static_dir() -> Path:
    """Absolute path to the static web UI directory served at ``/``."""
    return resource_root() / "src" / "static"


def default_log_dir() -> Path | None:
    """Fallback log directory when ``LOG_DIR`` is not set.

    - From source: None — logs go to the console.
    - Frozen: a ``logs`` directory beside the executable. The one-dir bundle is
      unpacked under the user's app-data directory, so it is writable. The
      Electron shell normally overrides this with ``LOG_DIR``.
    """
    if getattr(sys, "frozen", False):
        return Path(sys.executable).resolve().parent / "logs"
    return None
